src/tlo/bitset_handler/bitset_extension.py

```diff
@@ -0,0 +1,730 @@
+from __future__ import annotations
+
+import operator
+import re
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Dict,
+    Iterable,
+    List,
+    Optional,
+    Sequence,
+    Set,
+    Tuple,
+    Type,
+    TypeAlias,
+)
+
+import numpy as np
+import pandas as pd
+from numpy.dtypes import BytesDType
+from numpy.typing import NDArray
+from pandas._typing import TakeIndexer, type_t
+from pandas.core.arrays.base import ExtensionArray
+from pandas.core.dtypes.base import ExtensionDtype
+
+if TYPE_CHECKING:
+    from pandas._typing import type_t
+
+# Assume nodes are strings, else we can't construct from string when passed the name!
+# We can likely get around this with some careful planning, but we'd have to figure out how
+# to pass type-metadata for the elements from inside the output of self.name, so that casting
+# was successful.
+ALLOWABLE_ELEMENT_TYPES = (str,)
+BooleanArray: TypeAlias = np.ndarray[bool]
+CastableForPandasOps: TypeAlias = (
+    "NodeType"
+    | Iterable["NodeType"]
+    | NDArray[np.uint8]
+    | NDArray[np.bytes_]
+    | "BitsetArray"
+)
+SingletonForPandasOps: TypeAlias = "NodeType" | Iterable["NodeType"]
+NodeType: TypeAlias = str
+
+
+class BitsetDtype(ExtensionDtype):
+    """
+    A Bitset is represented by a fixed-width string, whose characters are each a uint8.
+    Elements of the set map 1:1 to these characters.
+
+    If the elements set is indexed starting from 0, then:
+    - The quotient of these indices (modulo 8) is the character within the string that contains the bit representing the element,
+    - The remainder (modulo 8) is the index within said character that represents the element itself.
+
+    The element map takes an element of the bitset as a key, and returns a tuple whose first element is the
+    corresponding string-character index, and the latter the uint8 representation of the element within that
+    string character.
+    """
+    _element_map: Dict[NodeType, Tuple[int, np.uint8]]
+    _elements: Tuple[NodeType]
+    _index_map: Dict[Tuple[int, np.uint8], NodeType]
+    _metadata = ("_elements",)
+
+    @classmethod
+    def construct_array_type(cls) -> type_t[BitsetArray]:
+        return BitsetArray
+
+    @classmethod
+    def construct_from_string(cls, string: str) -> BitsetDtype:
+        """
+        Construct an instance of this class by passing in a string of the form
+        that str(<instance of this class>) produces.
+        
+        That is, given a string of the form
+        bitset(#elements): e1, e2, e3, ...
+
+        this method will return a BitsetDtype with elements e1, e2, e3, ... etc.
+        """
```

Comment:
> would be a better alternative anyway? Is there a strong reason to allow both formats?

Pandas is the reason. The `dtype` argument to `pd.Series` needs to be able to take a string as an acceptable value, and then the default test suite seems to imply (quite reasonably) that

```python
pd.Series(some_data, dtype=str(BitsetDtype(e1, e2, ..., en)))
```

should work. IE that the output of `BitsetDtype.__str__` should be an acceptable argument to `dtype` as well as an explicit instance of `BitsetDtype` itself.

My intention with `BitsetDtype.__str__` was to display

```
bitset(#elements): e1, e2, ..., en
```

so that `print` and `pd.Series` display correctly, whilst also providing the user with all the at-a-glance information they need (this thing is a bitset of n elements, here's an explicit list). Granted, we potentially don't need the element count, but if we have a lot of elements (or _long_ names for the elements, I think it's useful!).

As such, this means that whatever we choose to display here is what `pandas` expects to be passed to `dtype`. My implementation of `construct_from_string` actually allows us to leave off the `bitset(#elements)` part because I originally wrote it assuming the input could be identical to `__init__`, not knowing about this caveat until after getting the test suite to work!

TBH I'd like to continue being able to `construct_from_string` by just passing the elements, though I'm not attached to this. I will add the check that the number of elements is correct though! :sweat_smile:

Reply:
Thanks for the detailed explanation 😁 - that all seems reasonable then!